### PR TITLE
Killing Some ToDos

### PR DESCRIPTION
**Pull request by steelbrain**
This is still a work in progress.
- [x] Make a base bottom tab class
- [x] Register file and project tabs as html custom elements
- [x] Add status as an html custom element
- [x] Remove all the old views logic

__Note:__ This PR removes the textBuffer parameter to lint function. That is an API breaking change.

### CODE REVIEW

**Comment by steelbrain** [comment]
The old views have been completely removed and replaced with no ones. Until this point, there is **no** API breaking change.

**Comment by steelbrain** [comment]
I am thinking of doing the priority fields in a separate PR. so Yeah, it's ready for a review.
/cc @park9140 @k2b6s9j @AsaAyers @hd-deman @iam4x

File: package.json [hunk 1]
```diff
@@ -20,7 +20,7 @@
     },
     "linter-plus": {
       "versions": {
-        "0.1.0": "consumeLinterPlus"
+        "0.1.0": "consumeLinter"
```
Comment (AsaAyers) [inline package.json]: If you're going to make a [breaking API change](https://github.com/AtomLinter/linter-coffeelint/blob/master/lib/plus-linter.coffee#L31) you need to bump this version number (removing 2nd param). semver.org says this needs to go to 0.2.0

Comment (k2b6s9j) [inline package.json]: :+1:

**Comment by k2b6s9j** [comment]
If this is still open on Sunday I'll review it then. I was at a funeral today and just laid in bed the rest of the day watching TV. I'll be busy all day tomorrow, Friday, and Saturday at a competition so I won't be able to look through, test, or code much until then. Feel free to move on without me.

File: package.json (hunk 1 above)
Comment (steelbrain) [inline package.json]: Good Idea, but because linter-plus isn't released yet, I am not sure if I should support legacy code or just bump the version.

File: lib/editor-linter.coffee [hunk 2]
```diff
@@ -83,5 +31,48 @@ class EditorLinter
...
+    return unless @editor is @linter.activeEditor
+    return if @_lock(wasTriggeredOnChange)
+    @lint(true) unless wasTriggeredOnChange # Trigger onFly linters on save.
 
-module.exports = EditorLinter
+    scopes = @editor.scopeDescriptorForBufferPosition(@editor.getCursorBufferPosition()).scopes
+    scopes.push '*' # To allow global linters
+
+    Promise.all(@_lint(wasTriggeredOnChange, scopes)).then =>
+      @_lock(wasTriggeredOnChange, false)
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: Why go back to having to check and set lock in 3 places instead of just letting the lock wrap the promise?

Comment (steelbrain) [inline lib/editor-linter.coffee]: IMO This is better understandable than that.

File: lib/editor-linter.coffee [hunk 3]
```diff
@@ -83,5 +31,48 @@ class EditorLinter
...
+        if linter.scope is 'project' then @linter.messagesProject.set linter, results
+        else @messages.set linter, results
+
+        @emitter.emit 'did-update'
+        @linter.views.render() if @editor is @linter.activeEditor
+
+  # This method sets or gets the lock status of given type
+  _lock: (wasTriggeredOnChange, value) ->
+    key = wasTriggeredOnChange ? 'inProgressFly' : 'inProgress'
+    return @[key] if typeof value is 'undefined'
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: IMO, this is one of the places where a standard if would be more clear.

```CoffeeScript
if typeof value is 'undefined'
    return @[key] 
@[key] = value
```

Comment (steelbrain) [inline lib/editor-linter.coffee]: Agreed.

File: lib/editor-linter.coffee [hunk 4]
```diff
@@ -83,5 +31,48 @@ class EditorLinter
...
+    @[key] = value
+
+  # Checks the responses for any kind-of errors
+  @validateResults: (results) ->
+    if (not results) or results.constructor.name isnt 'Array'
+      throw new Error "Got invalid response from Linter, Type: #{typeof results}"
+    results.forEach (Result) ->
```
Comment (AsaAyers) [inline lib/editor-linter.coffee]: The standard is for this to be `result` as it isn't a constructor.

Comment (steelbrain) [inline lib/editor-linter.coffee]: Your comment was very confusing, Please explain.

File: lib/h.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+"use strict"
+// H = Helpers
+let Helpers = {
```
Comment (AsaAyers) [inline lib/h.js]: This file appears to be unused.

Comment (steelbrain) [inline lib/h.js]: Not really, it's used in `linter-views::_extractMessages`.
Coffeescript doesn't support iterating over generators. so I had to write this in JS.

Comment (steelbrain) [inline lib/h.js]: It's assigned as @h to linter to be shared across the env.

Comment (AsaAyers) [inline lib/h.js]: oh, I searched for `h.js` and this is the only result.

Since this isn't a constructor it should be `helpers`

Comment (AsaAyers) [inline lib/h.js]: Also, `h` doesn't really mean anything. This would be better to name it `helpers.js` so it's immediately clear what it's purpose is.

Comment (steelbrain) [inline lib/h.js]: done

File: lib/editor-linter.coffee (hunk 4 above)
Comment (AsaAyers) [inline lib/editor-linter.coffee]:
```CoffeeScript
results.forEach (result) ->
    unless result.type
        throw new Error "Missing type field on Linter Response, Got: #{Object.keys(results)}"
```

1. the variable `Result` should not be capitalized.
2. you're referencing `restults` in the throw but it should also be `result`

You changed this to a `.forEach` which adds extra overhead of creating this function and running it multiple times. Why not just leave it as the for loop that was here before?

Comment (steelbrain) [inline lib/editor-linter.coffee]: Although the overhead is unnoticeable, your point is still valid. Done.

File: lib/linter-views.coffee [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+  # this method is called on package deactivate
+  deactivate: ->
+    @panel.removeDecorations()
+    @panelWorkspace.destroy()
+    @bubble?.remove()
+
+  # This method is called in render, and classifies the messages according to scope
+  _extractMessages: (Gen, counts) ->
+    isProject = @scope is 'project'
```
Comment (AsaAyers) [inline lib/linter-views.coffee]: Nothing in this file changes the `@scope` value. I'm not sure if it's unused or some other object is reaching into here and changing it. If that's the case a setter would make it more clear and easier to find whatever is changing it.

Comment (steelbrain) [inline lib/linter-views.coffee]: The tabs in the views change it's value.

File: lib/linter-views.coffee [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    activeFile = @linter.activeEditor.getPath()
+    ToReturn = []
+    @linter.h.genValues(Gen).forEach (Entry) ->
+      # Entry === Array<Messages>
+      Entry.forEach (message) ->
+        if (not message.file and not isProject) or message.file is activeFile
```
Comment (AsaAyers) [inline lib/linter-views.coffee]: I think this could use a comment. I'm finding it hard to parse what the intention is behind this line.

Comment (steelbrain) [inline lib/linter-views.coffee]: done

File: lib/views/bottom-tab-file.coffee [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  initialize: (@linter) ->
+    BottomTab.prototype.initialize.call(this, "Current File") # document.registerElement destroys it prototype chain
+  onClick: ->
+    @linter.views.scope = 'file'
+    @linter.views.bottomTabProject.active = false
+    @active = true
+    @linter.views.panel.render(@linter.views.messages)
+
+module.exports = BottomTabFile = document.registerElement('linter-bottom-tab-file',
+  {prototype: BottomTabFile.prototype})
```
Comment (AsaAyers) [inline lib/views/bottom-tab-file.coffee]: Based on http://www.html5rocks.com/en/tutorials/webcomponents/customelements/

maybe `document.registerElement('linter-bottom-tab-file',  {prototype: Object.create(BottomTabFile.prototype)})` would maintain the prototype chain?

File: lib/views/bottom-tab.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
+"use strict";
```
Comment (AsaAyers) [inline lib/views/bottom-tab.js]: ES6 implies strict mode

File: lib/views/bottom-tab-file.coffee (hunk 8 above)
Comment (steelbrain) [inline lib/views/bottom-tab-file.coffee]: This works pretty well at the moment, what do you mean?

Comment (AsaAyers) [inline lib/views/bottom-tab-file.coffee]: BottomTab.prototype.initialize.call(this, "Current File") # document.registerElement destroys it prototype chain

That line definitely shouldn't have to exist. Maybe if you use `Object.create` here you can remove that line from `initialize` and just use `super`

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (steelbrain) [inline lib/views/bottom-tab.js]: ES6 `Classes`, `let` and `const` are not available outside of strict mode in v8. If you haven't observed already there's no "use babel" on the header of it, which means this is parsed directly.

File: lib/views/bottom-tab-file.coffee (hunk 8 above)
Comment (steelbrain) [inline lib/views/bottom-tab-file.coffee]: Gonna text it in a tmp file and see if it works.

Comment (steelbrain) [inline lib/views/bottom-tab-file.coffee]: I tested it in a local js file and it worked even without `Object.create`, I guess coffeescript's way of extending classes is buggy. https://gist.github.com/steelbrain/50199ef361333bc61c2d

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (AsaAyers) [inline lib/views/bottom-tab.js]: You are correct. In that case I recommend `"use babel";` as it both provides strict mode and provides almost all of ES6. It's going to be frustrating and surprising the first time someone goes to use an ES6 feature that isn't available and it doesn't work just because we aren't using babel here.

The compiler caches the results, so I wouldn't worry about that overhead. All of the CoffeeScript has to be transpiled and cached anyway.

Comment (steelbrain) [inline lib/views/bottom-tab.js]: I am not sure if `"use strict"` introduces any kind-of problem. and I don't see a point in making use of babel when we don't really need it. All the ES6 features this file needs are already natively implemented. :)

File: lib/views/bottom-tab-file.coffee (hunk 8 above)
Comment (AsaAyers) [inline lib/views/bottom-tab-file.coffee]: it is possible, but it seems unlikely that CoffeeScript extends things in a non-standard way.

Comment (steelbrain) [inline lib/views/bottom-tab-file.coffee]: :smiley: Anyways, you can write the file I posted above in coffeescript and see for yourself.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (AsaAyers) [inline lib/views/bottom-tab.js]: The point is you're using some bizarre hybrid of ES5 and ES6 and I think sooner or later it's going to bite someone. Keep it if you like, we'll see how it goes. ¯\\_(ツ)_/¯

File: lib/views/bottom-tab-file.coffee (hunk 8 above)
Comment (AsaAyers) [inline lib/views/bottom-tab-file.coffee]: Since you are a collaborator it would really help if your branches where here. I'm trying to get this branch checked out now so I can change it here and see if it breaks.

Comment (steelbrain) [inline lib/views/bottom-tab-file.coffee]: Great, let me know how it goes.

File: lib/linter-views.coffee (hunk 6 above)
Comment (AsaAyers) [inline lib/linter-views.coffee]: Can we get something somewhere that indicates this? You should be able to understand a module by reading it. It's also often considered bad practice to have other objects just reach in and change things like this.

File: lib/views/bottom-tab-file.coffee (hunk 8 above)
Comment (AsaAyers) [inline lib/views/bottom-tab-file.coffee]: > CoffeeScript's way of extending stuff is BAD

When you make broad statements like this please keep in mind that it's only your opinion and you often don't have the full context of why things are this way. CoffeeScript's classes work very well, as long as you use it as a class with the `new` keyword.

I could say the way you're using CoffeeScript's class is BAD.

File: lib/linter-views.coffee (hunk 6 above)
Comment (steelbrain) [inline lib/linter-views.coffee]: I am not sure what makes you think that, I mean a function to set a value of this would be redundant 'cause we can just set it directly. and I'll add a note now

Comment (AsaAyers) [inline lib/linter-views.coffee]: Read this file in isolation and you can see that `@scope` never changes, so this check is useless. If you have to read and understand the entire codebase to understand a file you've done something horribly wrong.

If we take the approach that because it's `@scope` and not `@_scope` that it's public and other things might mess with it, then we must assume the same of `@bubble`, and `@panelWorkspace` and everything else that gets attached to this object.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (iam4x) [inline lib/views/bottom-tab.js]: Is it possible to use `'use babel'` with coffescript syntax?

Comment (AsaAyers) [inline lib/views/bottom-tab.js]: No, babel compiles your code to ES6 => ES5 and CoffeeScript compiles CoffeeScript => ES5

Comment (steelbrain) [inline lib/views/bottom-tab.js]: @iam4x Your question :smiley:

File: lib/linter-views.coffee (hunk 6 above)
Comment (steelbrain) [inline lib/linter-views.coffee]: If reading the codebase to understand a file is horribly wrong, then I am not sure if there's anything right in this project :smiley: , anyways, I've added a note for it.

and moving stuff to underscore isn't really necessary as no other class manipulates any of it's property except for scope. Come on man, It's quite self-explanatory.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (steelbrain) [inline lib/views/bottom-tab.js]: __Note:__ @iam4x that is a js file

Comment (iam4x) [inline lib/views/bottom-tab.js]: Sorry didn't see, was surprised to see `'use babel'` thought the project was in coffescript :)

Comment (steelbrain) [inline lib/views/bottom-tab.js]: Why not? cofeescript has **no** way to iterate over generators. and it doesn't support getter/setters either.

File: lib/linter-views.coffee (hunk 6 above)
Comment (AsaAyers) [inline lib/linter-views.coffee]: Then what's the point of modules? Why not just make this one gigantic file?

The underscore is private my convention because CoffeeScript classes don't enforce anything being private. There are a few options here.

Assume everything is public
=====================

When looking at this file you can't make any predictions on how it might execute because other code might change everything.

Assume everything is private
======================

As long as the rest of the code is well behaved this is fine, but now when I examine this file I see that `@scope` isn't needed, so lets just remove it. I know that will break other code only because you said other stuff messes with it.

Introduce underscores to indicate things are private
=======================================

Now it becomes obvious that `@scope` is public because everything else has been made private.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (AsaAyers) [inline lib/views/bottom-tab.js]: @steelbrain ! We agree on something from the start! It's awesome!

Personally I'd rather everything be ES6, but since Atom is CoffeeScript by default I accept that many people will want to use CofeeScript.

File: lib/linter-views.coffee (hunk 6 above)
Comment (steelbrain) [inline lib/linter-views.coffee]: You're hilarious, but I guess you make a point, not everyone bothers to read the whole project to see how a file works. so Yeah, lets make everything else underscored.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (iam4x) [inline lib/views/bottom-tab.js]: I would love to have everything in ES6, I was pushing this idea at first :+1: 

But now it seems a very big change.

Comment (steelbrain) [inline lib/views/bottom-tab.js]: We all agree, Yayyyyy.

File: lib/linter-views.coffee (hunk 6 above)
Comment (steelbrain) [inline lib/linter-views.coffee]: Done

**Comment by steelbrain** [comment]
Okay, so to avoid that coffeescript weirdness in bottom tabs, I've rewritten them in JS. Also moved the panel rendering logic to panel.

File: lib/linter-views.coffee [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    @bottomTabFile.count = counts.file
+    @bottomTabProject.count = counts.project
+    @_bottomStatus.count = counts.project
+
+  # This method is called when we get the status-bar service
+  attachBottom: (statusBar) ->
+    statusBar.addLeftTile
+      item: @bottomTabFile,
+      priority: -1001
```
Comment (park9140) [inline lib/linter-views.coffee]: magic priority numbers should be named constants.

File: lib/views/bottom-status.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    this.classList.add('inline-block')
+
+    this.iconSpan = document.createElement('span')
+    this.iconSpan.classList.add('icon')
+    this.appendChild(this.iconSpan)
+
+    this.count = 0
+  }
+  set count(Value){
+    if(Value){
```
Comment (park9140) [inline lib/views/bottom-status.js]: set should just change the value, rendering logic should be broken out.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (park9140) [inline lib/views/bottom-tab.js]: I'll just leave this here, https://github.com/eventualbuddha/decaffeinate

File: lib/views/bubble.coffee [hunk 12]
```diff
@@ -1,16 +1,28 @@
...
+    MessageLine.initialize(message, false)
+    root.appendChild MessageLine
+    if message.trace and message.trace.length
+      message.trace.forEach (trace) ->
+        MessageLine = new Message()
+        MessageLine.initialize(trace, true)
+        root.appendChild MessageLine
+    root
 
   update: (point) ->
```
Comment (park9140) [inline lib/views/bubble.coffee]: Almost everything in this method belongs in the view's model.

File: lib/views/bubble.coffee [hunk 13]
```diff
@@ -21,11 +33,10 @@ class Bubble
       @bubble = textEditor.decorateMarker(marker, {
         type: 'overlay',
         position: 'tail',
-        item: BubbleView(@linter, message)
+        item: @render(message)
```
Comment (park9140) [inline lib/views/bubble.coffee]: This line should be passing the model of the view we intend to render into the decoration.  If it has been properly registered with the view provider the decorator will recognize it and build the correct view to display.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (AsaAyers) [inline lib/views/bottom-tab.js]: That sounds nice, except he made the mistake of using Redux. I maintain CoffeeLint and Redux is garbage and appears to be abandoned. 

It doesn't support breaking conditions onto multiple lines. I've never needed to look further than that.

```CoffeeScript
if foo and
    bar
  undefined
```

File: lib/views/panel.coffee [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    @innerHTML = ''
+    messages.forEach (message) =>
+      if @linter.views.scope is 'file'
+        return unless message.currentFile
+      if message.currentFile and message.position
+        p = message.position
+        range = [[p[0][0] - 1, p[0][1] - 1], [p[1][0] - 1, p[1][1]]]
+        marker = @linter.activeEditor.markBufferRange range, {invalidate: 'never'}
+        @decorations.push @linter.activeEditor.decorateMarker(
+          marker, type: 'line-number', class: 'line-number-' + message.type.toLowerCase()
```
Comment (park9140) [inline lib/views/panel.coffee]: Use string templating here `"line-number-#{message.type.toLowerCase()}"

File: lib/views/message.coffee [hunk 15]
```diff
@@ -0,0 +1,48 @@
+# This file is imported in views/panel
+
+class Message extends HTMLElement
+  initialize: (message, addPath) ->
+    @message = message
```
Comment (park9140) [inline lib/views/message.coffee]: this method needs to be broken up into one or more render methods. and Some of the data mapping logic needs to be in the viewModel for this view.

File: lib/views/bottom-tab.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+class BottomTab extends HTMLElement{
+  initialize(Content){
+    this._active = true
```
Comment (park9140) [inline lib/views/bottom-tab.js]: initialization of data that should be on the view's model, need to break out render method.

File: lib/linter-views.coffee (hunk 10 above)
Comment (steelbrain) [inline lib/linter-views.coffee]: No idea what you're talking about. Please do explain.

File: lib/views/bottom-status.js (hunk 11 above)
Comment (steelbrain) [inline lib/views/bottom-status.js]: This seemed like the best way to me, it is the rendering element, so I am not sure where the logic should be

File: lib/views/bubble.coffee (hunk 12 above)
Comment (steelbrain) [inline lib/views/bubble.coffee]: Which view are you referring to?

File: lib/views/bubble.coffee (hunk 13 above)
Comment (steelbrain) [inline lib/views/bubble.coffee]: and what does that mean in simple english :P

File: lib/views/bottom-tab.js (hunk 16 above)
Comment (steelbrain) [inline lib/views/bottom-tab.js]: __Note:__ There is no render method.

File: lib/linter-views.coffee (hunk 10 above)
Comment (iam4x) [inline lib/linter-views.coffee]:
```
MAGIC_PRIORITY = -1001

attachBottom: (statusBar) ->
  statusBar.addLeftTile
    item: @bottomTabFile
    priority: MAGIC_PRIORITY
```

?

Comment (steelbrain) [inline lib/linter-views.coffee]: while removing the comma looks legit, I am not sure why would we need to do that.

File: lib/views/bottom-status.js (hunk 11 above)
Comment (park9140) [inline lib/views/bottom-status.js]: SIngle responsibility applies to more than classes.  The responsibility of the setter is to set the value then hand off to a known mechanism for the rendering.  Ideally you would have a render count change method that does the rendering.

File: lib/views/bottom-tab.js (hunk 16 above)
Comment (park9140) [inline lib/views/bottom-tab.js]: And yet there is rendering happening in this initialization...

File: lib/views/bubble.coffee (hunk 13 above)
Comment (park9140) [inline lib/views/bubble.coffee]: Remember our conversation about using the ViewProvider?  It's the same as that.

File: lib/views/bottom-status.js (hunk 11 above)
Comment (steelbrain) [inline lib/views/bottom-status.js]: @park9140 that method does nothing but render the value, and a setter IMHO is better looking, I mean, imagine `Tab.count = 2` vs `Tab.renderCount(2)`
the first one is better looking, and does the same thing.

File: lib/views/bottom-tab.js (hunk 16 above)
Comment (steelbrain) [inline lib/views/bottom-tab.js]: Hahaha. We don't really need a render method here, 'cause we don't re-create any dom elements. We just change their innerHTML or add/remove classes. The initialize sets initial values to be manipulated by those methods.

File: lib/views/bottom-status.js (hunk 11 above)
Comment (AsaAyers) [inline lib/views/bottom-status.js]: @park9140 Are you talking about just separating the logic like this:

```js
set count(value){
  this.value = value
  this.render()
}
render() {
  // do the rendering here
}
```

or making it two different calls:

```js
tab.count = 2
tab.render()
```

We've been discussing this in the chatroom and now that I see it's a DOM element, you do tend to expect them to change immediately.

File: lib/views/bottom-status.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+      this.classList.remove('linter-success')
+      this.iconSpan.classList.remove('icon-check')
+
+      this.classList.add('linter-error')
+      this.iconSpan.classList.add('icon-x')
+
+      this.iconSpan.textContent = Value === 0 ? '1 Error' : `${Value} Errors`
```
Comment (AsaAyers) [inline lib/views/bottom-status.js]: 0 is a falsy value and won't pass `if (Value) {` above, so the `1 Error` branch of this is unreachable. Why would value 0 mean 1 error?

Comment (steelbrain) [inline lib/views/bottom-status.js]: Good catch!

File: lib/views/bottom-status.js (hunk 11 above)
Comment (park9140) [inline lib/views/bottom-status.js]: The values should be updated by the value changing. No problem there. My problem is that we have multiple distinct rendering changes un named happening here and we have them happening inline to this method rather than having straight forward actions being called like 'renderNoErrors' or 'renderErrorCount' being called based on the value.

File: lib/views/bottom-tab.js (hunk 9 above)
Comment (park9140) [inline lib/views/bottom-tab.js]: If you logic is that involved you should be breaking it down for clarity anyway.

File: lib/views/bubble.coffee (hunk 12 above)
Comment (park9140) [inline lib/views/bubble.coffee]: The bubble view, I'm finding that we have yet again drifted from having models for our views.

File: lib/views/bubble.coffee [hunk 18]
```diff
@@ -1,16 +1,28 @@
...
+  render: (message) ->
+    root = document.createElement 'div'
+    root.id = 'linter-inline'
+    MessageLine = new Message()
+    MessageLine.initialize(message, false)
+    root.appendChild MessageLine
+    if message.trace and message.trace.length
+      message.trace.forEach (trace) ->
+        MessageLine = new Message()
+        MessageLine.initialize(trace, true)
```
Comment (park9140) [inline lib/views/bubble.coffee]: creating the view should not render it.

File: lib/views/bubble.coffee [hunk 19]
```diff
@@ -1,16 +1,28 @@
...
+        root.appendChild MessageLine
```
Comment (park9140) [inline lib/views/bubble.coffee]: Appending it to the dom should.

File: lib/views/bubble.coffee [hunk 20]
```diff
@@ -1,16 +1,28 @@
...
     @bubble = null
+  render: (message) ->
+    root = document.createElement 'div'
+    root.id = 'linter-inline'
+    MessageLine = new Message()
+    MessageLine.initialize(message, false)
+    root.appendChild MessageLine
+    if message.trace and message.trace.length
+      message.trace.forEach (trace) ->
+        MessageLine = new Message()
```
Comment (park9140) [inline lib/views/bubble.coffee]: Simple views can be constructed like this but breaking it into an initialization step is unnecessary especially when you are just setting initial view state, that should then be rendered based on it's addition to the dom.

File: lib/views/bubble.coffee (hunk 19 above)
Comment (steelbrain) [inline lib/views/bubble.coffee]: Acknowledged, I'll use the `onAttached` custom element callback and see how it goes.

File: lib/views/bottom-tab.js (hunk 16 above)
Comment (park9140) [inline lib/views/bottom-tab.js]: Any modification to a DOM element is rendering.  None of this should happen until the view is needed to be added to the dom, and only updated if the view has been added to the dom.  Views should be light to create so that model code can trigger their creation without incurring the cost of rendering. 

Rendering like this should utilize the same logic that later updates use to update components of the view as well.  In the lines below you initialize your content to '0' but don't use the rendering logic for the count.

File: lib/views/panel.coffee [hunk 21]
```diff
@@ -0,0 +1,45 @@
+Message = require './message'
```
Comment (park9140) [inline lib/views/panel.coffee]: We should probably rename all the views to 'xxxElement' to be clear they are views when they are being utilized.

File: lib/views/panel.coffee [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+      if @linter.views.scope is 'file'
+        return unless message.currentFile
+      if message.currentFile and message.position
+        p = message.position
+        range = [[p[0][0] - 1, p[0][1] - 1], [p[1][0] - 1, p[1][1]]]
+        marker = @linter.activeEditor.markBufferRange range, {invalidate: 'never'}
+        @decorations.push @linter.activeEditor.decorateMarker(
+          marker, type: 'line-number', class: "line-number-#{message.type.toLowerCase()}"
+        )
+        @decorations.push @linter.activeEditor.decorateMarker(
```
Comment (park9140) [inline lib/views/panel.coffee]: activeEditor is a model and as such should only be modified from a model.

File: lib/linter-views.coffee (hunk 10 above)
Comment (iam4x) [inline lib/linter-views.coffee]: It's for comprehension of code, when new people will read the code they won't try to search why there's `-1001` value provided.

**Comment by steelbrain** [comment]
Merging since there's not been any non-fixed known bugs and the view <--> model split would make this complicated. Gonna do it in another PR.
